consensus/core/src/core.rs

```diff
@@ -111,46 +156,207 @@ impl Core {
+
+        assert!(!to_propose.is_empty());
+
+        // Now clean up the pending ancestors
+        self.pending_ancestors
+            .retain(|round, _blocks| *round >= clock_round);
+
+        to_propose.into_iter().collect()
+    }
+
+    /// Checks whether all the leaders of the previous quorum exist.
+    /// TODO: we can leverage some additional signal here in order to more cleverly manipulate later the leader timeout
+    /// Ex if we already have one leader - the first in order - we might don't want to wait as much.
+    fn last_quorum_leaders_exist(&self) -> bool {
+        // TODO: check that we are ready to produce a new block. This will mainly check that the leaders of the previous
+        // quorum exist.
+        let quorum_round = self.threshold_clock.get_round().saturating_sub(1);
+
+        let leaders = self.leaders(quorum_round);
+        if let Some(ancestors) = self.pending_ancestors.get(&quorum_round) {
+            // Search for all the leaders. If at least one is not found, then return false.
+            // A linear search should be fine here as the set of elements is not expected to be small enough and more sophisticated
+            // data structures might not give us much here.
+            return leaders.iter().all(|leader| {
+                ancestors
+                    .iter()
+                    .any(|entry| entry.reference.author == *leader)
+            });
+        }
+        false
+    }
+
+    /// Returns the leaders of the provided round.
+    fn leaders(&self, round: Round) -> Vec<AuthorityIndex> {
+        // TODO: this info will be retrieved from the base committers. For now just do a simple round robin so we can
+        // use it in tests.
+        vec![AuthorityIndex::new_for_test(
+            round % self.context.committee.size() as u32,
+        )]
     }
 
     fn last_proposed_round(&self) -> Round {
-        self.last_own_block.round()
+        self.last_proposed_block.round()
+    }
+
+    fn last_proposed_block(&self) -> &VerifiedBlock {
+        &self.last_proposed_block
+    }
+
+    fn transactions_to_propose(&mut self) -> Vec<Transaction> {
+        let mut total_payload_size = 0;
+        let transactions_index = self
+            .pending_transactions
+            .iter()
+            .position(|t| {
+                total_payload_size += t.data().len();
+                total_payload_size > self.options.max_payload_size_bytes
+            })
+            .unwrap_or(self.pending_transactions.len());
+
+        let mut payload = self.pending_transactions.split_off(transactions_index);
+        mem::swap(&mut payload, &mut self.pending_transactions);
+
+        payload
+    }
+}
+
+/// A helper struct to keep a much reduced info set for a block that is pending to potentially be included
+/// in a next block proposal.
+struct PendingAncestorEntry {
+    reference: BlockRef,
+    timestamp: BlockTimestampMs,
+    ancestors: Vec<BlockRef>,
+}
```

Comment:
I would happily use the `VerifiedBlock` directly in the `pending_ancestors` map as in theory it should be cheap to copy (pretty much only a reference to the underlying data would be created) . @mwtian @arun-koshy if you think that this would be ok then I can refactor. Otherwise we can keep this reduced set which _would be ok_ to hold in memory anyways as the `pending_ancestors` map is quite ephemeral

Reply:
Yeah I think `VerifiedBlock` is cheaper to copy and store than the underlying fields, even just a subset of them. But `PendingAncestorEntry` should be cheap to move too. Also using `VerifiedBlock` seems simpler in not having to create a new type.

Reply:
I agree, using `VerifiedBlock` makes sense to me as well

Reply:
Ok I'll switch to using the `VerifiedBlock` and if this is proven to be expensive during testing then we can come back to the `PendingAncestorEntry` solution.